### ISSUE TITLE
Accuracy issues with TensorRT 8.6 when running Conformers on GPU A10G

### ISSUE THREAD

**Issue by buddhapuneeth** [NVIDIA/TensorRT#3081]
## Description
Conformer model on TensorRT with same set of input matrices is giving different results in each iteration.
This behavior leading to different final evaluation result for every run. 

I am conducting experiments with a batch size of 1 and a sequence length of 110. I am performing 110 rounds of evaluations in a recurrent manner. During these evaluations, I am capturing the model outputs for both correct and incorrect cases. The absolute differences between the outputs in these two scenarios are as follows:
`{"count":140800.0,"mean":0.0004827203,"std":0.0004537383,"min":0.0,"25%":0.000169085,"50%":0.000368,"75%":0.0006633,"max":0.010281}`

 I am observing variations in the results even within a single run, where the audio data has a sequence length of 110. To compare the outputs of different runs, I am utilizing the NVIDIA Polygraphy tool. Currently, I have exposed all matmul outputs (approximately 220) as graph outputs. Upon analysis, I have noticed slight deviations from the expected results, starting from the early matmul operation (around e-05). Based on my intuition, I believe that the accumulation of these small differences throughout the computation contributes to the variations observed in the final output. Hence, it is not solely attributed to one specific operator, but rather a cumulative effect.

Here is how I am using the Polygraphy tool:
```
from polygraphy.backend.onnxrt import OnnxrtRunner, SessionFromOnnx
from polygraphy.backend.trt import TrtRunner, Profile, CreateConfig, EngineFromNetwork, NetworkFromOnnxPath
from polygraphy.comparator import Comparator, DataLoader
import numpy as np

model_path = "cfr.onnx"

profiles = [
        Profile().add('input_114', min=[1, 1, 4608], opt=[2, 1, 4608], max=[2, 1, 4608])
                  .add(...) ...)
    ]

build_engine_32a = EngineFromNetwork(NetworkFromOnnxPath(model_path), config=CreateConfig(profiles=profiles))
build_engine_32b = EngineFromNetwork(NetworkFromOnnxPath(model_path), config=CreateConfig(profiles=profiles))

runners = [
    TrtRunner(build_engine_32a),
    TrtRunner(build_engine_32b),
]

data_seq = []
data = {}
# build the map of input_names and numpy_tensors
data_seq.append(data)
run_results = Comparator.run(runners, data_seq)

assert bool(Comparator.compare_accuracy(run_results))
run_results.save("inference_results.json")
```
Basically creating two runners for TensorRT and comparing results.
In this case, as I mentioned above one of the early matmul output is deferring at `1e-5` in the first iteration itself. 
A glimpse of that result from polygraphy tool is as below:

 ```
[I]     Comparing Output: 'cfr/tf.linalg.matmul_2/MatMul:0' (dtype=float32, shape=(1, 1, 1536)) with 'cfr/tf.linalg.matmul_2/MatMul:0' (dtype=float32, shape=(1, 1, 1536))
[I]         Tolerance: [abs=1e-05, rel=1e-05] | Checking elemwise error
[I]         trt-runner-N0-06/22/23-18:08:01: cfr/tf.linalg.matmul_2/MatMul:0 | Stats: mean=0.31248, std-dev=25.217, var=635.9, median=-0.85173, min=-169.14 at (0, 0, 523), max=295.77 at (0, 0, 1268), avg-magnitude=17.237
[I]             ---- Histogram ----
                Bin Range      |  Num Elems | Visualization
                (-169 , -123 ) |          1 |
                (-123 , -76.2) |          5 |
                (-76.2, -29.7) |        105 | ###
                (-29.7, 16.8 ) |       1116 | ########################################
                (16.8 , 63.3 ) |        295 | ##########
                (63.3 , 110  ) |          9 |
                (110  , 156  ) |          2 |
                (156  , 203  ) |          1 |
                (203  , 249  ) |          0 |
                (249  , 296  ) |          2 |
[I]         trt-runner-N1-06/22/23-18:08:01: cfr/tf.linalg.matmul_2/MatMul:0 | Stats: mean=0.31248, std-dev=25.217, var=635.9, median=-0.85172, min=-169.14 at (0, 0, 523), max=295.77 at (0, 0, 1268), avg-magnitude=17.237
[I]             ---- Histogram ----
                Bin Range      |  Num Elems | Visualization
                (-169 , -123 ) |          1 |
                (-123 , -76.2) |          5 |
                (-76.2, -29.7) |        105 | ###
                (-29.7, 16.8 ) |       1116 | ########################################
                (16.8 , 63.3 ) |        295 | ##########
                (63.3 , 110  ) |          9 |
                (110  , 156  ) |          2 |
                (156  , 203  ) |          1 |
                (203  , 249  ) |          0 |
                (249  , 296  ) |          2 |
[I]         Error Metrics: cfr/tf.linalg.matmul_2/MatMul:0
[I]             Minimum Required Tolerance: elemwise error | [abs=7.6294e-05] OR [rel=0.00046108] (requirements may be lower if both abs/rel tolerances are set)
[I]             Absolute Difference | Stats: mean=4.5774e-06, std-dev=4.8905e-06, var=2.3917e-11, median=3.8147e-06, min=0 at (0, 0, 2), max=7.6294e-05 at (0, 0, 523), avg-magnitude=4.5774e-06
[I]                 ---- Histogram ----
                    Bin Range            |  Num Elems | Visualization
                    (0       , 7.63e-06) |       1220 | ########################################
                    (7.63e-06, 1.53e-05) |        263 | ########
                    (1.53e-05, 2.29e-05) |         39 | #
                    (2.29e-05, 3.05e-05) |          7 |
                    (3.05e-05, 3.81e-05) |          3 |
                    (3.81e-05, 4.58e-05) |          0 |
                    (4.58e-05, 5.34e-05) |          2 |
                    (5.34e-05, 6.1e-05 ) |          0 |
                    (6.1e-05 , 6.87e-05) |          1 |
                    (6.87e-05, 7.63e-05) |          1 |
[I]             Relative Difference | Stats: mean=1.3819e-06, std-dev=1.5179e-05, var=2.3041e-10, median=2.6769e-07, min=0 at (0, 0, 2), max=0.00046108 at (0, 0, 1395), avg-magnitude=1.3819e-06
[I]                 ---- Histogram ----
                    Bin Range            |  Num Elems | Visualization
                    (0       , 4.61e-05) |       1532 | ########################################
                    (4.61e-05, 9.22e-05) |          1 |
                    (9.22e-05, 0.000138) |          1 |
                    (0.000138, 0.000184) |          0 |
                    (0.000184, 0.000231) |          0 |
                    (0.000231, 0.000277) |          0 |
                    (0.000277, 0.000323) |          0 |
                    (0.000323, 0.000369) |          1 |
                    (0.000369, 0.000415) |          0 |
                    (0.000415, 0.000461) |          1 |
[E]         FAILED | Output: 'cfr/tf.linalg.matmul_2/MatMul:0' | Difference exceeds tolerance (rel=1e-05, abs=1e-05)
```

Also without any change in polygraphy params, when I re-run the above script, I am seeing different numbers of failures.

I am aware of the precision issues with fp operations, but here surprisingly results at some position has a gap of 0.010281.

## Environment
**TensorRT Version**: 8.6

**NVIDIA GPU**: A10G

**NVIDIA Driver Version**: 515.65.01  

**CUDA Version**: 11.7

Operating System: AL2

API: C++

Precision: TF32

## Steps To Reproduce
 Unfortunetely, I cannot provide the source model. But the model has below operations:
 `Less,Div,Shape,Concat,Sigmoid,Tanh,Cos,Cast,Not,Pad,BatchNormalization,ReduceProd,And,Sub,Gather,MatMul,Reciprocal,Greater,Range,Transpose,Add,Max,Expand,GlobalAveragePool,Squeeze,ReduceMean,Reshape,Sin,Slice,Equal,Mul,Unsqueeze,Sqrt,ReduceSum,Tile,LessOrEqual,Split,ReduceSumSquare,Softmax`

 Also, during build phase, I noticed this warning around reshape operation and I believe it wasn't the reason for this issue.
```

[W] Tensordot: IShuffleLayer with zeroIsPlaceHolder=true has reshape dimension at position 1 that might or might not be zero. TensorRT resolves it at runtime, but this may cause excessive memory consumption and is usually a sign of a bug in the network.
[W] Profile kMIN values are not self-consistent. IShuffleLayer Tensordot: reshaping failed for tensor: Tensordot/MatMul:0 reshape would change volume 0 to 262144 Instruction: RESHAPE_ZERO_IS_PLACEHOLDER{0 512} {1 0 512}.
```

**Have you tried [the latest release](https://developer.nvidia.com/tensorrt)?**: yes

**Comment by zerollzeng** [comment]
Do you mean with an engine, feeds the same input for every iteration, the output is different?

**Comment by zerollzeng** [comment]
Have you compare the e2e accuracy with onnx runtime? e.g. `polygraphy run model.onnx --trt --onnxrt`

**Comment by buddhapuneeth** [comment]
Do you mean with an engine, feeds the same input for every iteration, the output is different?

> Yes, same model and same inputs on same setup giving slightly different results, as I mentioned above sometime the difference in outputs for two runs is:
`{"count":140800.0,"mean":0.0004827203,"std":0.0004537383,"min":0.0,"25%":0.000169085,"50%":0.000368,"75%":0.0006633,"max":0.010281}`
But here in this case, I am building model again for each iteration from ONNX. May be the build strategy is causing this issue. I am planning to verify it and let you know.

Have you compare the e2e accuracy with onnx runtime? e.g. polygraphy run model.onnx --trt --onnxrt

> I compared the output of the ORT equivalent CPU-based engine outside the Polygraphy tool and observed significant precision differences (e-03) with certain GPU results during some iterations. However, some iterations of GPU results were sufficiently close to the CPU results (e-05). The issue I want to emphasize here is the variation in GPU results across iterations, which I am able to reproduce using the Polygraphy tool as mentioned in the initial description.

**Comment by Slyne** [comment]
@buddhapuneeth 
Please help me understand your question better. 
1. You use the same tensorrt engine (not same onnx model) and feed in different input, but the result is different. 
2. You made comparison between output from onnx model running on CPU and the generated tensorrt engine running on GPU, but even the input is the same,  the result is different on some iterations?  Could you share the codes/scripts for doing the comparison?  Maybe you can check if you add the synchronize() to wait for GPU to complete all the computing.

**Comment by zerollzeng** [comment]
Would be great if you can provide the onnx for reproduce.

**Comment by ttyio** [comment]
closing since no activity for more than 3 weeks, pls reopen if you still have question, thanks!

**Comment by LinuxCup** [comment]
@buddhapuneeth ,Have you resolved? I encountered the similar problem. the result of versions of 8.5 and 8.6 is inconsistent, running the engine file that from the same onnx file.
